### DISCUSSION TITLE
Approach circle looks blurred with FR mod on

### DISCUSSION THREAD

**Discussion by telecomadm1145** [ppy/osu#23110] [Q&A] [answered]
### Type

Cosmetic

### Bug description

When I enable the freeze frame mod in osu!, I notice that the edges of the approach circle become blurred. (looks ugly) Upon examining the code (osu.game.ruleset.osu and osu.game), I discovered that both the legacy and default skins use the same texture component, which has a resolution of only 256x256. I believe this is causing the issue.

To fix the problem, I've been trying to replace the approach circle component in the skin transformer with a RingPiece component. However, I've found that the beatmaps' legacy skin always returns its original approach circle component, regardless of my changes. As a result, my patch only seems to work when the "beatmap skins" option is turned off.

I'm currently using version 2023.403.1 of osu! with the d3d11 renderer, and a modified version based on an old clone (although I'm not sure which version it is).

### Screenshots or videos

![2023.403.1](https://user-images.githubusercontent.com/47657299/229434958-21612ad6-7164-4802-b262-c1a33b6fd86d.png)
2023.403.1 d3d11 (blurred edge)
![patch](https://user-images.githubusercontent.com/47657299/229435372-9465384b-ec4c-46d0-9ba7-929c81f06d24.png)
This is my patch, which is actually my work-in-progress skin for Lazer. I apologize that the realm of development version storage is located in another folder.
![patch_2](https://user-images.githubusercontent.com/47657299/229435737-b11a53bd-2854-41f2-804b-429d5ea62f66.png)
Here is my patch with "beatmap skins" turned on, showing the blurred edge.

### Version

2023.403.1

### Logs

actually theres nothing for uploading.
2023.403.1
[network.log](https://github.com/ppy/osu/files/11135684/network.log)
[performance.log](https://github.com/ppy/osu/files/11135685/performance.log)
[updater.log](https://github.com/ppy/osu/files/11135687/updater.log)
[runtime.log](https://github.com/ppy/osu/files/11135686/runtime.log)
[database.log](https://github.com/ppy/osu/files/11135688/database.log)
[legacy-ipc.log](https://github.com/ppy/osu/files/11135689/legacy-ipc.log)
my patch
[input.log](https://github.com/ppy/osu/files/11135697/input.log)
[performance.log](https://github.com/ppy/osu/files/11135702/performance.log)
[runtime.log](https://github.com/ppy/osu/files/11135703/runtime.log)
[database.log](https://github.com/ppy/osu/files/11135704/database.log)
[legacy-ipc.log](https://github.com/ppy/osu/files/11135700/legacy-ipc.log)

**Comment by peppy** [comment, marked as answer]
A good starting point would be to update the approach circle only for the default skins.

One thing to keep in mind is that the fact the border thickness gets thicker when the circle is larger is an unfortunate side effect of using sprites until now. I'd argue the border should remain the same (in fact, original ouendan did this) and that should be a potentially breaking change, pending user feedback.
